### PR TITLE
CS-829 Save the Call Recording Link for the Phone Call

### PR DESCRIPTION
**Pull request by rogeryang-sugarcrm**

### CODE REVIEW

**Review by jwelborn-sugar — CHANGES_REQUESTED** [review]

File: src/utils/aws/s3-utils.js [hunk 1]
```diff
@@ -41,4 +41,35 @@ async function getJsonFromS3Event(event) {
...
 }
 
-module.exports = { getJsonFromS3Event };
+/**
+ * Get the object key from the S3 trigger event
+ *
+ * @param {Object} event S3 trigger event
+ */
+function getObjectKeyFromS3Event (event) {
+    return event.Records[0].s3.object.key;
```
[the remark this answers is not in the document]
Comment (rogeryang-sugarcrm) [reply src/utils/aws/s3-utils.js]: Good point! Added a decode to all keys

File: src/utils/sugar/call-record-utils.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  // return empty if AWS Connect instance name is not defined
+  if (!connectInstance) {
+    return '';
+  }
+
+  let url = UrlConstants.HTTPS_PREFIX + connectInstance + UrlConstants.AWS_CONNECT_URL_PARTIAL + 
+    UrlConstants.AWS_CONNECT_CALL_RECORDING_URL_PARTIAL + contactId;
+
+  return url;
```
[the remark this answers is not in the document]
Comment (rogeryang-sugarcrm) [reply src/utils/sugar/call-record-utils.js:65]: The URL partials are already URL safe, but we can be extra safe by encoding the custom parameters

**Review by jwelborn-sugar — DISMISSED** [review]

File: src/utils/sugar/call-record-utils.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sugarhqi) [reply src/utils/sugar/call-record-utils.js:65]: If I understand correctly, we only use this function to construct a url. Why do we need a lambda function and this PR when the same thing can be done on sugar side easily with all the data available in sugar: instance name (from config) and contactId (saved in call record)?

**Review by sugarhqi — CHANGES_REQUESTED** [review]

**Comment by rogeryang-sugarcrm** [comment]
@sugarhqi, for some reason, I can't directly respond to your comment above so I will do it here

The purpose of the Lambda function is to let Sugar know that some call record has a call recording available. We don't know if the contact flow has call recording set up, so we can use a Lambda function that watches the S3 bucket. If we add a call recording link to every call, some links may break if they do not have an audio file

From the Lambda side, we can either send the URL or a flag indicating that there is an available recording

File: src/utils/sugar/call-record-utils.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (rogeryang-sugarcrm) [reply src/utils/sugar/call-record-utils.js:65]: Per the [comment](https://sugarcrm.atlassian.net/browse/CS-829?focusedCommentId=797202) on CS-829, we have decided that the Lambda function will generate the URL and send it to Sugar

**Review by sugarhqi — APPROVED** [review]

**Pull request merged**